bootloader/lkenv/lkenv.go

```diff
@@ -0,0 +1,278 @@
+}
+
+func (l *Env) Set(key, value string) {
+	switch key {
+	case "snap_mode":
+		copyString(l.env.Snap_mode[:], value)
+	case "snap_kernel":
+		copyString(l.env.Snap_kernel[:], value)
+	case "snap_try_kernel":
+		copyString(l.env.Snap_try_kernel[:], value)
+	case "snap_core":
+		copyString(l.env.Snap_core[:], value)
+	case "snap_try_core":
+		copyString(l.env.Snap_try_core[:], value)
+	}
+}
+
+// Configure partition labels for used boot partitions
+// this should not be used at run time, it should be used
+// once at image build time, if part labels are not filled by gadget build
+func (l *Env) ConfigureBootPartitions(boot_1, boot_2 string) {
+	copyString(l.env.Bootimg_matrix[0][0][:], boot_1)
+	copyString(l.env.Bootimg_matrix[1][0][:], boot_2)
+}
+
+func (l *Env) Load() error {
+	err := l.LoadEnv(l.path)
+	if (err != nil) {
+		return l.LoadEnv(l.pathbak)
+	}
+	return nil
+}
+
+func (l *Env) LoadEnv(path string) error {
+	f, err := os.Open(path)
+	if err != nil {
+		return fmt.Errorf("error loading lk environment file: %v", err)
+	}
+
+	defer f.Close()
+	if err := binary.Read(f, binary.LittleEndian, &l.env); err != nil {
+		return fmt.Errorf("error reading lk environment from file %v", err)
+	}
+	// calculate crc32 to validate structure
+	w := bytes.NewBuffer(nil)
+	ss := binary.Size(l.env)
+	w.Grow(ss)
+	err = binary.Write(w, binary.LittleEndian, &l.env)
+	if err != nil {
+		return fmt.Errorf("error writting lk environment to buffer for validation %v", err)
+	}
+	crc := crc32.ChecksumIEEE(w.Bytes()[:ss-4]) // size of crc32 itself at the end of the structure
+	if crc != l.env.Crc32 {
+		return fmt.Errorf("lkenv:load: broken crc32 [0x%X]vs[0x%X]\n", crc, l.env.Crc32)
+	}
+	logger.Debugf("Load: validated crc32 (0x%X)", l.env.Crc32)
+	return nil
+}
+
+func (l *Env) Save() error {
+	logger.Debugf("Save")
+	w := bytes.NewBuffer(nil)
+	ss := binary.Size(l.env)
+	w.Grow(ss)
+	err := binary.Write(w, binary.LittleEndian, &l.env)
+	if err != nil {
+		return fmt.Errorf("error writting lk environment to buffer for saving %v", err)
+	}
+	// calculate crc32
+	l.env.Crc32 = crc32.ChecksumIEEE(w.Bytes()[:ss-4])
+	logger.Debugf("Save: calculated crc32 (0x%X)", l.env.Crc32)
+	w.Reset()
+	err = binary.Write(w, binary.LittleEndian, &l.env)
+	if err != nil {
+		return fmt.Errorf("error re-writting lk environment to buffer for saving %v", err)
+	}
+	err = l.SaveEnv(l.path, w)
+	if err != nil {
+		logger.Debugf("Save: failed to save main environment")
+	}
+	// if there is backup environment file save to it as well
+	if(osutil.FileExists(l.pathbak)) {
+		errr := l.SaveEnv(l.pathbak, w)
+		if errr != nil {
+			logger.Debugf("Save: failed to save backup environment")
+		}
+	}
+	return err;
+}
+
+func (l *Env) SaveEnv(path string, buf *bytes.Buffer) error {
+	f, err := os.OpenFile(path, os.O_WRONLY, 0660)
+	if err != nil {
+		return fmt.Errorf("error opening lk env file for env storing %v", err)
+	}
+	defer f.Close()
+
+	if _, err := f.Write(buf.Bytes()); err != nil {
```

Comment:
and then io.Copy(f, buf)

Reply:
explained in previous why this does not work